### ISSUE TITLE
cmake module: integrate external library as subproject

### ISSUE THREAD

**Issue by falk-werner** [mesonbuild/meson#7244]
**tl;dr**

This is a feature request: I would like to integrate an external cmake project as a subproject using a combination of meson's cmake module and a wrap file. 

**Long version**

There is this library, [libwebsockets](https://libwebsockets.org/), that I want to use in my project. When the library is already installed in the system, I want to use the it from the system. Otherwise I want meson to download it and use it as a subproject.

If the library were a meson project, I would have used the dependency function and a wrap file:

````
lws_dep = dependency('libwebsockets', fallback: ['libwebsockets', 'libwebsockets_dep'])
````

Since libwebsockets is a CMake project, it won't be that easy. Fortunately, I've learned, that there is [WrapDB](https://mesonbuild.com/Using-the-WrapDB.html) with lots of existing wraps for popular libraries. There was no existing wrap for the library, so I've created one and everything is fine.

While writing the wrap (the library comes with a 3.000 lines CMakeLists.txt and it took a while to write the corresponding meson.build file) I became unsure whether this is the right solution. 
I learned about the cmake module and started to combine it with a wrap file. It seems to work and there is a small example here: [https://github.com/falk-werner/meson-evaluation/tree/master/cmake_sub](https://github.com/falk-werner/meson-evaluation/tree/master/cmake_sub)

A little confused, I've [asked](https://github.com/mesonbuild/libwebsockets/pull/1#issuecomment-637736321) if this solution is okay and what's about all the existing wraps of popular cmake projects like gtest, GSL, jansson etc.. The answer I got says, that the wrap system was not designed for my use case and there might be some problems when I'm using it that way.

I think, my use case is not that uncommon and it would be great to get official support for this.

_(Otherwise I have to rework this very long [meson.build](https://github.com/mesonbuild/libwebsockets/blob/4.0.13/meson.build) file every time a new version of libwebsockets is released or somehow convince the maintainer to support meson himself.)_

**Comment by mensinda** [comment]
There is the [cmake module](https://mesonbuild.com/CMake-module.html) which implements CMake subprojects. 

Note: See the [mixing build systems disclaimer](https://mesonbuild.com/Mixing-build-systems.html) and using the latest meson (0.54.2) and CMake (3.17) is **highly** recommended.

Do you have any feedback regarding the docs, since it looks like you missed this feature?

**Comment by mensinda** [comment]
Oh, sorry ignore my last comment I didn't read your post completely...

But still, what you are describing should already work out of the box, even your cmake_sub works for me and looks totally fine (and is the recommended way to do things). Could you please clarify your problem?

**Comment by falk-werner** [comment]
My cmake_sub works also for me. 
But I wonder why there are so many wraps for popular cmake projects in the WrapDB, when this can be solved by using the cmale module.

I asked this quesion already here:
[https://github.com/mesonbuild/libwebsockets/pull/1#issuecomment-637736321](https://github.com/mesonbuild/libwebsockets/pull/1#issuecomment-637736321)

The answer was, that the cmake module was not designed for this use case.

**Comment by falk-werner** [comment]
Sorry, ... the answer was that the wrap system was not designed for this use case.
If you say, the way I'm using the cmake module in my cmake_sub is fine, the issue can be closed.

**Comment by mensinda** [comment]
The CMake module *is* designed to work with the wrap system (as in wrap files, not WrapDB). Not sure where this myth came from. You should only be aware of the [official disclaimer](https://mesonbuild.com/Mixing-build-systems.html) and always use the latest meson and CMake versions.

Also, WrapDB is much older than CMake subprojects. This is why there are some popular projects on WrapDB.

**Comment by falk-werner** [comment]
Thanks for the clarification. This helps a lot.
Sorry for any inconveniences.

**Comment by mensinda** [comment]
No problem.
